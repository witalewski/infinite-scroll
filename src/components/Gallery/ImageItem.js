import React, { Component } from 'react';
import { observer, inject } from 'mobx-react';
import styled from '@emotion/styled';

const ImageItemStyled = styled.div`
  position: relative;
  width: 100%;
  height: 100%;

  .image {
    opacity: 0;
    position: absolute;
  }

  .image--loaded {
    opacity: 1;
    transition: opacity 0.5s;
  }

  .overlay {
    position: absolute;
    top: 50%;
    width: 100%;
    height: 40%;

    display: flex;
    justify-content: center;
    align-items: center;

    opacity: 1;

    background: rgba(255, 255, 255, 0.9);

    transition: opacity 0.2s;
    transform: translateY(-50%);
  }

  .overlay--hidden {
    opacity: 0;
  }

  .primary-button--hidden {
    display: none;
  }

  .message {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  .message--hidden {
    display: none;
  }

  .favourites-indicator {
    position: absolute;
    bottom: 12px;
    right: 12px;
    transition: opacity 0.5s 0.2s;
  }

  .favourites-indicator--hidden {
    opacity: 0;
  }
`;

export class ImageItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isImageLoaded: false,
      mouseOver: false,
      message: '',
      primaryButtonVisible: false,
      imageHeight: 0,
    };
  }

  componentWillUnmount() {
      if (this.timer) {
          clearTimeout(this.timer);
      }
  }
  onImageLoad = ({ target }) => {
    this.setState({ isImageLoaded: true , imageHeight: target.clientHeight});
  };

  onMouseOver = () => {
    this.setState({ mouseOver: true });
  };

  onMouseOut = () => {
    if (!this.state.message) {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
      });
    }
  };

  addToFavourites = () => {
    this.props.addToFavourites(this.props.image);
    this.setState({
      message: 'Added to favourites!',
      undoAction: this.removeFromFavourites,
    });
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
        primaryButtonVisible: false,
      });
    }, 2000);
  };

  removeFromFavourites = () => {
    this.props.removeFromFavourites(this.props.image);
    this.setState({
      message: 'Removed from favourites.',
      undoAction: this.addToFavourites,
    });
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
        primaryButtonVisible: false,
      });
    }, 2000);
  };

  render() {
    const { image, favourites } = this.props;
    const { mouseOver, isImageLoaded, imageHeight } = this.state;
    const isFavourite = favourites.includes(this.props.image);
    return (
      <ImageItemStyled
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
        style={{height: imageHeight || this.props.height}}
      >
        {image && (
          <img
            className={`image ${isImageLoaded && 'image--loaded'}`}
            alt="Shibe dog"
            src={image}
            decoding="async"
            onLoad={this.onImageLoad}
          />
        )}
        <div className={`overlay ${!mouseOver && 'overlay--hidden'}`}>
          <button
            type="button"
            className={`btn btn-outline-primary primary-button ${this.state
              .message && 'primary-button--hidden'}`}
            onClick={
              isFavourite ? this.removeFromFavourites : this.addToFavourites
            }
          >
            {isFavourite ? 'üíî Remove from Favourites' : '‚ù§Ô∏è Add to Favourites'}
          </button>
          <div
            className={`message ${!this.state.message && 'message--hidden'}`}
          >
            <span>{this.state.message}</span>
            <button
              type="button"
              className="btn btn-secondary btn-sm"
              onClick={this.state.undoAction}
            >
              Undo
            </button>
          </div>
        </div>
        <div
          className={`favourites-indicator ${!isFavourite &&
            'favourites-indicator--hidden'}`}
        >
          <span role="img" aria-label="favourite">
            ‚ù§Ô∏è
          </span>
          Ô∏è
        </div>
      </ImageItemStyled>
    );
  }
}

export default inject(({ appState }) => ({
  favourites: appState.favourites,
  addToFavourites: appState.addToFavourites,
  removeFromFavourites: appState.removeFromFavourites,
}))(observer(ImageItem));
