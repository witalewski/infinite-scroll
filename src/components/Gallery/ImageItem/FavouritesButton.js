import React from 'react';
import styled from '@emotion/styled';

const FavouritesButtonStyled = styled.button`
  position: absolute;

  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  width: 120px;
  height: 120px;
  padding-top: 12px;
  font-size: 64px;

  background: rgba(255, 255, 255, 0.8);
  border: none;

  transition: opacity 0.5s;

  &.btn--hidden {
    opacity: 0;
  }
`;

export const FavouritesButton = ({ visible, isFavourite, onButtonClick }) => (
  <FavouritesButtonStyled
    className={`btn btn-light ${!visible && 'btn--hidden'}`}
    type="button"

  >
    {isFavourite ? '💔' : '❤️'}
  </FavouritesButtonStyled>
);
