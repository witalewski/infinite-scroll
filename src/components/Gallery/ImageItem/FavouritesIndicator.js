import React from 'react';
import styled from '@emotion/styled';

const FavouritesIndicatorStyled = styled.div`
  position: absolute;
  bottom: 12px;
  right: 12px;
  transition: opacity 0.5s 0.2s;

  &.favourites-indicator--hidden {
    opacity: 0;
  }
`;
export const FavouritesIndicator = ({ isFavourite }) => (
  <FavouritesIndicatorStyled
    className={`favourites-indicator ${!isFavourite &&
      'favourites-indicator--hidden'}`}
  >
    <span role="img" aria-label="favourite">
      ❤️
    </span>
  </FavouritesIndicatorStyled>
);
