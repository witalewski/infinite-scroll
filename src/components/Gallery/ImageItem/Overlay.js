import React from 'react';
import styled from '@emotion/styled';

const OverlayStyled = styled.div`
    position: absolute;
    top: 50%;
    width: 100%;
    height: 40%;

    display: flex;
    justify-content: center;
    align-items: center;

    opacity: 1;

    background: rgba(255, 255, 255, 0.9);

    transition: opacity 0.2s;
    transform: translateY(-50%);

  &.overlay--hidden {
    opacity: 0;
  }

  .primary-button--hidden {
    display: none;
  }

  .message {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }
`;

export const Overlay = ({
  mouseOver,
  message,
  isFavourite,
  addToFavourites,
  removeFromFavourites,
  undoAction,
}) => (
  <OverlayStyled className={`overlay ${!mouseOver && 'overlay--hidden'}`}>
    <button
      type="button"
      className={`btn btn-outline-primary primary-button ${message &&
        'primary-button--hidden'}`}
      onClick={isFavourite ? removeFromFavourites : addToFavourites}
    >
      {isFavourite ? '💔 Remove from Favourites' : '❤️ Add to Favourites'}
    </button>
    { message && (<div className="message">
      <span>{message}</span>
      <button
        type="button"
        className="btn btn-secondary btn-sm"
        onClick={undoAction}
      >
        Undo
      </button>
    </div>)}
  </OverlayStyled>
);
