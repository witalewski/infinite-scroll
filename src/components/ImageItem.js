import React, { Component } from 'react';
import styled from '@emotion/styled';

const ImageItemStyled = styled.div`


  }
  .image {
    opacity: 0;
  }

  .image--loaded {
    opacity: 1;
    transition: opacity 0.5s;
  }
`;

export class ImageItem extends Component {
  onImageLoad({ target }) {
    target.classList.add('image--loaded');
    target.parentNode.parentNode.style.height = `${target.clientHeight}px`;
  }

  render() {
    const { image } = this.props;
    return (
      <ImageItemStyled>
        {image.isPlaceholder ? (
          <div className="image-placeholder">
          </div>
        ) : (
          <img
            className="image"
            alt="Shibe dog"
            src={image.url}
            decoding="async"
            onLoad={this.onImageLoad}
          />
        )}
      </ImageItemStyled>
    );
  }
}
