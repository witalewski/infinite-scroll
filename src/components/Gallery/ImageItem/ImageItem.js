import React, { Component } from 'react';
import { observer, inject } from 'mobx-react';
import styled from '@emotion/styled';
import { FavouritesIndicator } from './FavouritesIndicator';
import { FavouritesButton } from './FavouritesButton';

const ImageItemStyled = styled.div`
  position: relative;
  width: 100%;
  height: 100%;

  cursor: pointer;

  .image {
    opacity: 0;
    position: absolute;
  }

  .image--loaded {
    opacity: 1;
    transition: opacity 0.5s;
  }
`;

export class ImageItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isImageLoaded: false,
      mouseOver: false,
      imageHeight: 0,
    };
  }
  onImageLoad = ({ target }) => {
    this.setState({ isImageLoaded: true, imageHeight: target.clientHeight });
  };

  onMouseOver = () => {
    this.setState({ mouseOver: true });
  };

  onMouseOut = () => {
    if (!this.state.message) {
      this.setState({
        mouseOver: false,
      });
    }
  };

  addCurrentImageToFavourites = () => {
    this.props.addToFavourites(this.props.image);
  };

  removeCurrentImageFromFavourites = () => {
    this.props.removeFromFavourites(this.props.image);
  };

  render() {
    const { image, favourites } = this.props;
    const { mouseOver, isImageLoaded, imageHeight } = this.state;
    const isFavourite = favourites.includes(this.props.image);
    return (
      <ImageItemStyled
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
        style={{ height: imageHeight || this.props.height }}
        onClick={
          isFavourite
            ? this.removeCurrentImageFromFavourites
            : this.addCurrentImageToFavourites
        }
      >
        {image && (
          <img
            className={`image ${isImageLoaded && 'image--loaded'}`}
            alt="Shibe dog"
            src={image}
            decoding="async"
            onLoad={this.onImageLoad}
          />
        )}
        <FavouritesButton visible={mouseOver} isFavourite={isFavourite} />
        <FavouritesIndicator isFavourite={isFavourite} />
      </ImageItemStyled>
    );
  }
}

export default inject(({ appState }) => ({
  favourites: appState.favourites,
  addToFavourites: appState.addToFavourites,
  removeFromFavourites: appState.removeFromFavourites,
}))(observer(ImageItem));
