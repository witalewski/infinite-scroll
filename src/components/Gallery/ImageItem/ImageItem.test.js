import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Set } from 'immutable';
import mockImageURLs from '../../../fixtures/imageURLs.json';
import { ImageItem } from './ImageItem';

describe('ImageItem', () => {
  it('matches snapshot for favourite image', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <ImageItem
        image={mockImageURLs[0]}
        height={400}
        favourites={new Set(mockImageURLs)}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  it('matches snapshot for non-favourite image', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <ImageItem
        image={mockImageURLs[0]}
        height={400}
        favourites={new Set([])}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
