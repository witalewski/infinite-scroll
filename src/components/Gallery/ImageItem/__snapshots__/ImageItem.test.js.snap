// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageItem matches snapshot for favourite image 1`] = `
<ForwardRef(render)
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "height": 400,
    }
  }
>
  <img
    alt="Shibe dog"
    className="image false"
    decoding="async"
    onLoad={[Function]}
    src="https://cdn.shibe.online/shibes/8047547e25c8cdd6bf562e4c42dfa5afb9e8bdd6.jpg"
  />
  <FavouritesButton
    isFavourite={true}
    visible={false}
  />
  <FavouritesIndicator
    isFavourite={true}
  />
</ForwardRef(render)>
`;

exports[`ImageItem matches snapshot for non-favourite image 1`] = `
<ForwardRef(render)
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "height": 400,
    }
  }
>
  <img
    alt="Shibe dog"
    className="image false"
    decoding="async"
    onLoad={[Function]}
    src="https://cdn.shibe.online/shibes/8047547e25c8cdd6bf562e4c42dfa5afb9e8bdd6.jpg"
  />
  <FavouritesButton
    isFavourite={false}
    visible={false}
  />
  <FavouritesIndicator
    isFavourite={false}
  />
</ForwardRef(render)>
`;
